### PR TITLE
Add state store to electron main process

### LINKED ISSUE #1930: saveAs dialog is able to be triggered when no notebook is open
Steps to reproduce:

* Run nteract on macOS
* Close all nteract windows
* Click "File -> Save As"

### PR DESCRIPTION
**Pull request by captainsafia**
Closes #1930 (and a couple of others).

`store` is stored in a global variable (*gasp*). This should be moved to a closure.

We might want to move other things into the main process store. For example, the GitHub token.

Also I should add tests.

But this is a start.

### CODE REVIEW

File: packages/desktop/src/main/menu.js [hunk 1]
```diff
@@ -585,23 +249,371 @@ export function loadFullMenu(kernelSpecs) {
...
         click: createSender("menu:interrupt-kernel")
       },
       {
         label: "Restart Running Kernel",
+        enabled: BrowserWindow.getFocusedWindow() !== null,
         click: createSender("menu:restart-kernel")
       },
       {
         label: "Restart and Clear All Cells",
+        enabled: BrowserWindow.getFocusedWindow() !== null,
```
Comment (rgbkrk) [inline packages/desktop/src/main/menu.js:616]: Sharp! I see why this has to be redone on `focus`, `show`, etc.

**Review by rgbkrk — APPROVED** [review]
The cool thing about this at least is that at some point we'll be able to use this setup to update the kernelspecs so that users don't have to restart to get their available kernelspecs.

**Comment by rgbkrk** [comment]
> store is stored in a global variable (gasp). This should be moved to a closure.

That's definitely ok for now, we can leave it for a refactor. We may want to leave a `TODO: ` or `HACK: ` comment for now.

File: packages/desktop/src/main/index.js [hunk 2]
```diff
@@ -217,7 +222,8 @@ fullAppReady$.subscribe(() => {
   kernelSpecsPromise
     .then(kernelSpecs => {
       if (Object.keys(kernelSpecs).length !== 0) {
-        const menu = loadFullMenu(kernelSpecs);
+        store.dispatch(setKernelSpecs(kernelSpecs));
```
Comment (rgbkrk) [inline packages/desktop/src/main/index.js:226]: I look forward to us reacting to new kernelspecs and changing the Menu as a result. I wish our API (internally) for this was:

```
<Menu>
  {bunchaMenus}
  <ViewMenu />
  <KernelMenu kernelspecs={kernelspecs} />
</Menu>
```

File: packages/desktop/src/main/reducers.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
+// @flow
+
```
Comment (rgbkrk) [inline packages/desktop/src/main/reducers.js:2]: 👍

File: packages/desktop/src/main/menu.js [hunk 4]
```diff
@@ -552,25 +231,10 @@ export const named = {
   ]
 };
 
-export function generateDefaultTemplate() {
-  const template = [];
-
-  if (process.platform === "darwin") {
-    template.push(named);
-  }
```
Comment (rgbkrk) [inline packages/desktop/src/main/menu.js:560]: One way you could make this testable is by accepting the store as a variable and letting it default to using the global store:

```js
export function loadFullMenu(store=global.store) {
  ...
}
```

Or exporting a bound and unbound version.

**Comment by rgbkrk** [comment]
Thanks for tackling this one, it's definitely a yucky one!

File: packages/desktop/src/main/menu.js (hunk 4 above)
Comment (captainsafia) [reply packages/desktop/src/main/menu.js:560]: Sounds good! I'll add this change in the PR with the test cases.

**Comment by captainsafia** [comment]
I'm gonna merge this for now and add a testing setup for main process reducers in an upcoming PR.

**Comment by rgbkrk** [comment]
Awesome, thanks!

**Pull request merged**